[Xu BoJie] iP

bojie3 opened the pull request:
# Duke CLI
> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

Duke CLI is useful as it is
* Text-based
* Easy to learn and use
* Fast

And it is **Free**

Features of Duke CLI:
1. Add Task
2. Delete Task 
3. Mark and Unmark Task as done
4. Find tasks

The main function to run is in Duke.java as shown below
```java
    public static void main(String[] args) {
        Duke duke = new Duke();
        duke.run();
    }
```
A Duke object is created and `public void run()` is called on the duke object.


Future development plan 📈 
- [x] Add ability to store Task
- [ ] Include the ability to search for tasks based on time
- [ ] Add the choice to include locations into tasks

Commit:
Add Gradle support

bojie3 committed:
Level 1 finished

bojie3 committed:
Finished level 2

bojie3 committed:
Finished level 3

bojie3 committed:
Changed to level 2 to add tag

bojie3 committed:
Finished to level 3

bojie3 committed:
Finished level 4

bojie3 committed:
Finished A-TextUiTesting

bojie3 committed:
Fixed minor bug in task name and updated testcases

bojie3 committed:
Finished implementation for level-5

bojie3 committed:
Finished level 6

bojie3 committed:
Added storage, ui, command and parser class

bojie3 committed:
Added DateTime class to manage dates and time

bojie3 committed:
Merge branch 'branch-Level-7'
Include level 7 (Add storage, Ui, Parser class) into the main branch.

bojie3 committed:
Added task to data.txt to test storage.java

bojie3 committed:
Merged branch-level-8 to master to add datetime class to master branch

bojie3 committed:
Added Duke package and its subpackages

bojie3 committed:
Merge branch 'add-gradle-support' of https://github.com/bojie3/ip
Merge to include gradle support

bojie3 committed:
Added gradle support

bojie3 committed:
Added JUnit tests cases for Todo and DukeException

bojie3 committed:
Updated build.grade mainClassName to create jar file

bojie3 committed:
Added javadocs for classes and public methods

bojie3 committed:
Update all packages to lowercase and capitalise constants

bojie3 committed:
Add a temp folder to move files

bojie3 committed:
Delete folder Duke because of wrong capitalisation

bojie3 committed:
Delete temp folder and add duke folder with correct capitalisation

bojie3 committed:
Update command classes to be nouns

bojie3 committed:
Add feature to find task

bojie3 committed:
Merge branch 'branch-A-JavaDoc'
Update master branch to include JavaDocs

bojie3 committed:
Update master branch to include updated coding standards

bojie3 committed:
Update master branch to include find task function

bojie3 committed:
Update master branch to fix errors in naming of folders

bojie3 committed:
Add JavaDocs for function supporting find feature

bojie3 committed:
Update name of boolean variable according to standards

bojie3 committed:
Update code to fix checkstyle errors

bojie3 committed:
Add GUI feature

bojie3 committed:
Merge branch 'branch-Level-10' Add GUI feature to main branch

bojie3 committed:
Update code to remove CLI functions

bojie3 committed:
Merge branch 'branch-A-Varargs' To include changes in the optional use
of Varargs

bojie3 committed:
Add javadocs for GUI classes and functions

mayuanxin1234 commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Class in charge of handling the case of adding Deadline Task
+ */
+public class AddDeadlineCommand extends Command {
```
I like how you have a Command class that branches out to the different commands for duke respectively which follows good OOP principles.

mayuanxin1234 commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,75 @@
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
+ */
+public class DialogBox extends HBox {
```
This might be more suited to be in the Ui class?

mayuanxin1234 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,283 @@
+     * @return Command corresponding to the input of user
+     * @throws DukeException If user input is unknown or incorrect format
+     */
+    public Command parse(String str) throws DukeException {
```
Can explore using enums and switch to make the code look cleaner and shorter.

mayuanxin1234 commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,23 @@
+     * @return String version of todotask
+     */
+    @Override
+    public String toString() {
```
Can simplify to using ? and : instead of if else to make code look simplier.

mayuanxin1234 reviewed (commented):
Looks good! Well written code

JavonTeo commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+ * A class for managing tasks
+ */
+public class Duke{
+    private Ui ui;
```
Since ui is not changed at all throughout the class, perhaps it could be a constant instead?

JavonTeo commented on src/main/java/duke/command/DeleteTaskCommand.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public String execute(TaskList tl, Ui ui, Storage storage) {
+        storage.delete(index);
+        Task t = tl.getTask(index - 1);
```
Would task t be ambiguous naming? Perhaps a more obvious naming would be "task".

JavonTeo commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,155 @@
+    private static final String INTRO = "Hello! I'm Duke\nWhat can I do for you?\n";
+    private static final String OUTRO = "Bye. Hope to see you again soon!\n";
+
+    public Ui() {}
```
As empty constructors do not need to be declared, perhaps it would be a good idea to remove this constructor?

JavonTeo reviewed (commented):
LGTM almost, just some nits to consider.

bojie3 committed:
Add assert statement in Duke

bojie3 committed:
Add assert statement to getResponse function

Add assert since user input should not be null under any circumstances.
Update the javaDocs for getResponse function

bojie3 committed:
Update variabe t to task in deleteCommandTask

Changed the variable naming to imporve code quality.
Change variable t (supposed to represent a task) to task

bojie3 committed:
Merge pull request #1 from bojie3/branch-A-Assertions

Add assert statement

bojie3 committed:
Merge pull request #2 from bojie3/branch-A-CodeQuality

Update variabe t to task in deleteCommandTask

bojie3 committed:
Add function to sort the tasks according to their end time

Sorts the tasks and output the sorted tasks to the user.
Also updates the task in the data.txt to be sorted.
Place event/deadline < todo, and between event and deadline, we compare
the ending time of event to the deadline of deadline task.

bojie3 committed:
Add javadocs for sortCommand class

bojie3 committed:
Update JavaDocs and fix Checkstyle errors

bojie3 committed:
Update README.md

bojie3 committed:
Update README.md

bojie3 committed:
Update README.md

bojie3 committed:
Add png file

bojie3 committed:
Merge branch 'master' of https://github.com/bojie3/ip
Pull code to include updates in readme

bojie3 committed:
Update javadocs for functions

bojie3 committed:
Add title to GUI and update Ui.png

bojie3 committed:
Update README.md

bojie3 committed:
Update README.md
